### PR TITLE
feat(tracer): Support for external observability providers - Tracer

### PR DESCRIPTION
**Pull request by roger-zhangg**
**Issue number:**
[2030](https://github.com/awslabs/aws-lambda-powertools-python/issues/2030)

## Summary
Support provider in tracer
### Changes
Modify the provider with more genreral tracer feature, provide a new x-ray provider.

### User experience
This change doesn't affect users who just use default x-ray provider.
This change allows users to bring their own tracer provider to use other solution provider other than X-ray

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by roger-zhangg** [comment]
Drafted the first PR for tracer provider.
- Provided a more general Tracer Provider interface.
- Drafted a x-ray provider
I wasn't included in the initial discussions in this projects so I might have missed something.

Questions:
- Decide how to refactor `BaseSegment` Class
- Check on current x-ray provider's design.(Doesn't support `add_span`, `end_span`)
- In the RFC there's a capture_method function included in the provider. But currently this function is in Tracer class. Do we need to move it into our provider?

**Comment by leandrodamascena** [comment]
Hey @rubenfonseca! Could you please review this PR? As we discussed offline, here are some key points to consider during the review:

1 - DD and OTEL are just placeholders/examples in the code and will be removed before merging this PR.
2 - We've updated naming conventions to use more conventional names in the methods.
3 - Regarding the example, would it be better to create a new one instead of modifying the existing example?

Thanks for your attention to these points.

File: aws_lambda_powertools/tracing/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/base.py]:
```suggestion
## MAINTENANCE: deprecated, kept for backwards compatibility until v3 is released
```

And can you please create a new issue to track the removal of this class on V3?

File: aws_lambda_powertools/tracing/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/base.py]: This needs to be addressed on the RFC. I have a hunch that instrumentation needs nanoseconds, but we need to make sure this will be the common denominator amongst all providers.

File: aws_lambda_powertools/tracing/provider/__init__.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/__init__.py]: Note: in the future, maybe we should avoid this, as loading one provider will probably load all of them:

`from aws_lambda_powertools.tracing.provider import XrayProvider`

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: I can't find any reference in the RFC why a `number.Number` is used here. Can you give me more context?

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]:
```suggestion
        """Set an attribute for a span with a key-value pair.
```

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]:
```suggestion
        """Records an exception.
```

I'm sorry but "to trace entities" means nothing in my mind :( Can it maybe be rephrased?

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: I don't understand this line.

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: Can we get a bigger description about what a "Span" is?

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py:37]: Can we have string docs for this class please?

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: I need much more than this line of documentation. I know already what it returns based on the signature. I'm more interested about what it does and when it should be used.

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]:
```suggestion
            Optional parameters to be propagated to the span
```

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: I need much more than this line of documentation. I know already what it returns based on the signature. I'm more interested about what it does and when it should be used.

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]:
```suggestion
            Optional parameters to be propagated to the span
```

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: Is "trace entity" the "current span" here?

File: aws_lambda_powertools/tracing/provider/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/base.py]: I need more explanation here, why a string? How can libraries be passed on to this method? Can we have an example on the string doc?

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]: I'm trusting you did your research here, as I don't have enough knowledge to help here :D

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]: Again, I understand that you are removing the other providers; however, the manner in which you have written the code results in importing all providers despite utilizing only one of them.

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]: I believe since no `kwargs` are being passed here, all  "attributes" will be treated as metadata, and never as annotation. Is this the correct behaviour?

```python
    def set_attribute(self, key: str, value: str | Number | bool, **kwargs) -> None:
        if kwargs.get("namespace", "") != "":
            self.put_metadata(key=key, value=value, namespace=kwargs["namespace"])
        else:
            self.put_annotation(key=key, value=value)
```

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]: What if you just use `set_attribute` and make sure it works with both annotations and metadata on the Xray provider?

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]: What if you just use `set_attribute` and make sure it works with both annotations and metadata on the Xray provider?

File: examples/tracer/src/sdk_escape_hatch.py
Comment (rubenfonseca) [inline examples/tracer/src/sdk_escape_hatch.py]: Please keep the old example as it was to make sure it still works and maybe create a new one?

File: poetry.toml
Comment (rubenfonseca) [inline poetry.toml]: Are these changes intended on this PR?

File: pyproject.toml
Comment (rubenfonseca) [inline pyproject.toml]: Probably needs to be moved to a dev dependency.

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]: I don't understand this code. If I'm passing another provider that is not an instance of the BaseProvider, it will use the XRayProvider, is that correct? Is this intendend?

Overall I would prefer to remove this method, and fix this in the concrete implementation. This would remove a lot of the `# type:ignore` errors that leaves me scared.

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py:739]: BaseSpan is not compatible with BaseSegment, so I would say this is a breaking change. BUT since it's a "private" method, I think it's ok. Just wanted you to double check it's fine.

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
Thank you for the PR, there's a lot of great work here! 

I've left some suggestions. Can I also asked to update the PR body with a better description of the changes? It took me a while going back and forth to understand what was the scope of the PR.

File: aws_lambda_powertools/tracing/base.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/base.py]: Created #4105

File: aws_lambda_powertools/tracing/base.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/base.py]: @leandrodamascena and I have a discussion on this and we come to the conclusion that using float in epoch seconds this case is more appropriate. I also created a [issue](https://github.com/aws/aws-xray-sdk-python/issues/424)  for this in X-Ray python SDK as I believe the docstring here is referencing the X-Ray Python SDK. Will make sure to mention this in RFC

File: aws_lambda_powertools/tracing/provider/__init__.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/provider/__init__.py]: removed

File: aws_lambda_powertools/tracing/provider/base.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/provider/base.py]: Changed to Any to remove limitation.

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]: need further discussion, see [here](https://github.com/aws-powertools/powertools-lambda-python/pull/2342#discussion_r1560159107)

File: aws_lambda_powertools/tracing/tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/tracer.py]: removed

File: aws_lambda_powertools/tracing/tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/tracer.py]: changed the behavior of x_ray provider's `set_attribute`. Now namespace will be passed in.

File: aws_lambda_powertools/tracing/tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/tracer.py]: This is to make sure the Tracer works the same way as it works before on X-Ray provider. We have considered unifying the behavior using `set_attribute`. 

To do so, firstly we need to make sure all the existing code calling `Tracer.put_annotation` is still calling `put_annotation` in SDK. But we didn't figure out a good way to do that. 

Secondly, current unit test is testing extensively if `provider.put_annotation`/`provider.put_metadata` is called. We need to change a lot of test cases to change `put_annotation` call `provider.set_attribute` instead. I think we will need an additional discussion to signoff the final design on this.

File: aws_lambda_powertools/tracing/tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/tracer.py]: same as above

File: aws_lambda_powertools/tracing/tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/tracer.py:739]: Yes this is tested.

File: poetry.toml
Comment (roger-zhangg) [reply poetry.toml]: remove before merge

File: pyproject.toml
Comment (roger-zhangg) [reply pyproject.toml]: remove before merge

File: aws_lambda_powertools/tracing/tracer.py
Comment (roger-zhangg) [reply aws_lambda_powertools/tracing/tracer.py]: The reason for this method to exist is to keep the current x-ray tracer and test work as intended. this `_is_custom_provider` is used only to decide if we should call `provider.set_attribute` or the original `provider.put_annotation`/`provider.put_metadata` in Tracer `put_annotation`, `put_metadata` methods. 

The idea is when we introduce custom provider in the future, these two method: `put_annotation`, `put_metadata` should call `provider.set_attribute` instead. But for X-Ray provider and the test dummy provider will get `false` from `_is_custom_provider` and uses current `provider.put_annotation`/`provider.put_metadata`. 

As for your suggestion: " I would prefer to remove this method, and fix this in the concrete implementation" I do agree. To do so, we need to make `provider.set_attribute` to work for both `put_annotation`, `put_metadata` in X-ray sdk and also fix all the impacted test cases (for example [here](https://github.com/aws-powertools/powertools-lambda-python/blob/0ac38003c860749cb9d0baef8630de7eba1c8733/tests/unit/test_tracing.py#L141-L160)). My thinking before is we should avoid modifying test cases for this feature.

**Comment by roger-zhangg** [comment]
@rubenfonseca Thanks for the super detailed feedback. I've resolved all the comments now. Please help with a final review, thanks.

File: examples/tracer/src/sdk_escape_hatch.py
Comment (rubenfonseca) [inline examples/tracer/src/sdk_escape_hatch.py:14]: We should review this to make sure we don't have a typing breaking change

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]:
```suggestion
        """Set an attribute on current active span with a key-value pair.
```

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]:
```suggestion
        key: str
```

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]:
```suggestion
        value: Any
```

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]:
```suggestion
        Set an attribute for a pseudo service named payment
```

File: aws_lambda_powertools/tracing/tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/tracer.py]: Due to the new code, this comment can probably be deleted

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]: I think we're trying to use full imports now

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        Set an attribute on this span with a key-value pair.
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        key: str
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        value: Any
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        category: Literal["Annotation","Metadata","Auto"] = "Auto"
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        Set an attribute on the current active span with a key-value pair.
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        key: str
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        value: Any
```

File: aws_lambda_powertools/tracing/provider/xray/xray_tracer.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/provider/xray/xray_tracer.py]:
```suggestion
        category: Literal["Annotation","Metadata","Auto"] = "Auto"
```

File: aws_lambda_powertools/tracing/base.py
Comment (rubenfonseca) [inline aws_lambda_powertools/tracing/base.py]: Is this a breaking change? Or are we assuming that no one is inheriting from `BaseSegment` today?

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
Super good progress! Just left some tiny picks and some questions on breaking changes. @leandrodamascena can you help implement the changes so we commit quickly?

File: aws_lambda_powertools/tracing/base.py
Comment (leandrodamascena) [reply aws_lambda_powertools/tracing/base.py]: We will deprecate this in Powertools v3. We're not using this class today, but we never know if customers are, so it's best to revert this change.

Reverted.

**Comment by leandrodamascena** [comment]
We can consider this PR completed, but we will not merge it now. We'll start conversations about Powertools V3 and add this refactor when we're ready for it:

1 - We will deprecate the [BaseSegment](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/tracing/base.py#L8) in Powertools v3.
2 - This is a [potential breaking change](https://github.com/aws-powertools/powertools-lambda-python/pull/2342/files#diff-ac2e7c32fce52e9e17243aafe7f69a62615e44d9d3a15dabd164d16b476dc5daR18) for customers who are [accessing the XRay](https://docs.powertools.aws.dev/lambda/python/latest/core/tracer/#escape-hatch-mechanism) methods directly.
3 - We can take this opportunity to standardize method names like `trace` instead of `subsegment` in our documentation and ease for customers adopt new Observability providers.

I'm approving this PR and changing it to Draft state just so I don't mess up our stats.

**Review by leandrodamascena — APPROVED** [review]
Thanks for another stellar job @roger-zhangg! This refactor will open the opportunity to expand the Tracer utility and add more observability providers.

I left a comment explaining why we are approving but not merging this PR right now: 
https://github.com/aws-powertools/powertools-lambda-python/pull/2342#issuecomment-2061734362

**Comment by leandrodamascena** [comment]
Hi @roger-zhangg! I'm closing this PR in favor of #4902! It was really hard to fix the entire merge conflict here, so I've created a new one.

Thank you for all the effort you've put in here, and I'll give you credit for this work when I release v3.

**Pull request closed without merging**
